File: src/calculate_winner.js
import winningIndexesCombinations from "./winning_combination_of_squares";

export default function calculateWinner(squares) {

  for (let i = 0; i < winningIndexesCombinations.length; i++) {

    const [a, b, c] = winningIndexesCombinations[i];

    if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
      return squares[a];
    }
    /*to prevent the condition from evaluating to true if the square at position a is empty (null).
    If the square at position a is empty, the condition will evaluate to false without checking the other squares.
    This prevents a false positive for a winner if one of the squares is empty*/

  }
  return null;
}

File: api/calculate_winner.py
from winning_combination_of_squares import winningIndexesCombinations

def calculateWinner(squares):
    for winningCombination in winningIndexesCombinations:
        a, b, c = winningCombination

        # if the same symbol exists in a group of three squares from one of the winning combination, we have a winner
        if squares[a] and squares[a] == squares[b] and squares[a] == squares[c]:
            return { 
               "winner": squares[a] 
            }
    return { "winner" : None }